daemon/response.go

```diff
@@ -281,55 +282,142 @@ func (f fileResponse) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 //
 // Tip: “jq” knows how to read this; “jq --seq” both reads and writes this.
 type journalLineReaderSeqResponse struct {
-	io.ReadCloser
-	follow bool
+	readers []io.ReadCloser
+	follow  bool
+}
+
+func (rr *journalLineReaderSeqResponse) logReader(r io.ReadCloser, c chan systemd.Log, e chan error) {
+	defer r.Close()
+	decoder := json.NewDecoder(r)
+	for {
+		var log systemd.Log
+
+		// This will always cause an error before or later because of an
+		// io.EOF. This means we can rely on this being our termination
+		// condition for the read loop, and then do the error handling in
+		// the main go routine.
+		if err := decoder.Decode(&log); err != nil {
+			e <- err
+			break
+		}
+		c <- log
+	}
 }
 
 func (rr *journalLineReaderSeqResponse) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	w.Header().Set("Content-Type", "application/json-seq")
 
 	flusher, hasFlusher := w.(http.Flusher)
-
-	var err error
-	dec := json.NewDecoder(rr)
 	writer := bufio.NewWriter(w)
 	enc := json.NewEncoder(writer)
-	for {
-		var log systemd.Log
-		if err = dec.Decode(&log); err != nil {
-			break
-		}
 
-		writer.WriteByte(0x1E) // RS -- see ascii(7), and RFC7464
+	// start all the readers
+	c := make(chan systemd.Log)
+	e := make(chan error)
```

Comment:
is intentional here to use unbuffered channel, so that everything waits for log lines to be consumed one by one?

Reply:
No that was actually not intentional! I had experimented a bit and had not switched to buffered channels, thank you for pointing this out. The code has been updated to use buffered channels.